File: Python/581. Shortest Unsorted Continuous Subarray.py
class Solution:
    def findUnsortedSubarray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        ans, l, r = 0, 0, len(nums) - 1
        while l < len(nums)-1 and nums[l] <= nums[l+1]:
            l += 1
        while r > 0 and nums[r] >= nums[r-1]:
            r -= 1
        if l < r:
            mi = min(nums[l:r+1])
            ma = max(nums[l:r+1])
            while l >= 0 and nums[l] > mi:
                l -= 1
            while r < len(nums) and nums[r] < ma:
                r += 1
            ans = r - l - 1
        return ans

File: C++/581. Shortest Unsorted Continuous Subarray.cpp
class Solution {
public:
    int findUnsortedSubarray(vector<int>& nums) {
        int ans = 0;
        int l = 0, r = nums.size()-1;
        while(l<nums.size()-1 && nums[l] <= nums[l+1])
            l += 1;
        while(r>0 && nums[r] >= nums[r-1])
            r -= 1;
        if(l < r)
        {
            int mi = nums[l], ma = nums[r];
            for(int i=l;i<=r;i++)
            {
                mi = min(mi,nums[i]);
                ma = max(ma,nums[i]);
            }
            while(l>=0 && nums[l] > mi)
                l--;
            while(r<nums.size() && nums[r] < ma)
                r++;
            return r-l-1;
        }
        return ans;
    }
};